File: usvm-ts/src/test/kotlin/org/usvm/samples/Async2.kt
package org.usvm.samples

import org.jacodb.ets.model.EtsScene
import org.junit.jupiter.api.Disabled
import org.usvm.api.TsTestValue
import org.usvm.util.TsMethodTestRunner
import kotlin.test.Test

class Async2 : TsMethodTestRunner() {

    companion object {
        private const val SDK_TYPESCRIPT_PATH = "/sdk/typescript"
        private const val SDK_OHOS_PATH = "/sdk/ohos"
    }

    private val className = this::class.simpleName!!

    override val scene: EtsScene = loadSampleScene(
        className,
        sdks = listOf(SDK_TYPESCRIPT_PATH, SDK_OHOS_PATH)
    )

    @Test
    fun `chain promises`() {
        val method = getMethod(className, "chainPromises")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 20.0 },
            )
        )
    }

    @Test
    fun `chain with rejection`() {
        val method = getMethod(className, "chainWithRejection")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 42.0 },
            )
        )
    }

    @Test
    fun `await async function`() {
        val method = getMethod(className, "awaitAsyncFunction")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 10.0 },
            )
        )
    }

    @Test
    fun `multiple awaits`() {
        val method = getMethod(className, "multipleAwaits")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 8.0 },
            )
        )
    }

    @Test
    fun `promise all`() {
        val method = getMethod(className, "promiseAll")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 6.0 },
            )
        )
    }

    @Test
    fun `promise race`() {
        val method = getMethod(className, "promiseRace")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 42.0 },
            )
        )
    }

    @Disabled("Long running failing test")
    @Test
    fun `nested async`() {
        val method = getMethod(className, "nestedAsync")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 20.0 },
            )
        )
    }

    @Test
    fun `conditional await true`() {
        val method = getMethod(className, "conditionalAwait")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 100.0 },
            invariants = arrayOf(
                { r -> r.number == 100.0 || r.number == 200.0 },
            )
        )
    }

    @Test
    fun `async try catch`() {
        val method = getMethod(className, "asyncTryCatch")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 50.0 },
            )
        )
    }

    @Test
    fun `async finally`() {
        val method = getMethod(className, "asyncFinally")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 15.0 },
            )
        )
    }

    @Test
    fun `concurrent operations`() {
        val method = getMethod(className, "concurrentOperations")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 35.0 },
            )
        )
    }

    @Test
    fun `immediate resolve`() {
        val method = getMethod(className, "immediateResolve")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 123.0 },
            )
        )
    }

    @Test
    fun `immediate reject`() {
        val method = getMethod(className, "immediateReject")
        discoverProperties<TsTestValue>(
            method = method,
            { r -> r is TsTestValue.TsException },
        )
    }

    @Test
    fun `conditional promise resolve`() {
        val method = getMethod(className, "conditionalPromise")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 777.0 },
            invariants = arrayOf(
                { r -> r.number == 777.0 },
            )
        )
    }

    @Test
    fun `conditional promise reject`() {
        val method = getMethod(className, "conditionalPromise")
        discoverProperties<TsTestValue>(
            method = method,
            { r -> r is TsTestValue.TsException },
        )
    }

    @Test
    fun `await primitives`() {
        val method = getMethod(className, "awaitPrimitives")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 42.0 },
            )
        )
    }

    @Test
    fun `mixed values`() {
        val method = getMethod(className, "mixedValues")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 60.0 },
            )
        )
    }

    @Test
    fun `recursive async`() {
        val method = getMethod(className, "recursiveAsync")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number >= 0.0 }, // Sum of 1+2+...+n should be positive
            invariants = arrayOf(
                { r -> r.number >= 0.0 },
            )
        )
    }

    @Test
    fun `nested promise resolution`() {
        val method = getMethod(className, "nestedPromiseResolution")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 88.0 },
            )
        )
    }

    @Test
    fun `error propagation`() {
        val method = getMethod(className, "errorPropagation")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            invariants = arrayOf(
                { r -> r.number == 99.0 },
            )
        )
    }
}

File: usvm-ts/src/test/resources/samples/Async2.ts
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols,ES6RedundantAwait

class Async2 {
    // Test chaining promises
    chainPromises(): number {
        const promise1 = Promise.resolve(10);
        const promise2 = promise1.then((value) => value * 2);
        return await promise2; // Should return 20
    }

    // Test promise chaining with rejection
    chainWithRejection(): number {
        const promise1 = Promise.reject(new Error("First error"));
        const promise2 = promise1.catch((error) => 42);
        return await promise2; // Should return 42
    }

    // Test async function returning a promise
    async asyncFunction(value: number): Promise<number> {
        if (value > 0) {
            return value * 2;
        } else {
            throw new Error("Value must be positive");
        }
    }

    // Test awaiting an async function
    awaitAsyncFunction(): number {
        return await this.asyncFunction(5); // Should return 10
    }

    // Test multiple awaits in sequence
    multipleAwaits(): number {
        const promise1 = Promise.resolve(5);
        const promise2 = Promise.resolve(3);
        const result1 = await promise1;
        const result2 = await promise2;
        return result1 + result2; // Should return 8
    }

    // Test Promise.all
    promiseAll(): number {
        const promises = [
            Promise.resolve(1),
            Promise.resolve(2),
            Promise.resolve(3),
        ];
        const results = await Promise.all(promises);
        return results.reduce((sum, val) => sum + val, 0); // Should return 6
    }

    // Test Promise.race
    promiseRace(): number {
        const promises = [
            Promise.resolve(42),
            Promise.resolve(100),
        ];
        return await Promise.race(promises); // Should return 42 (first resolved)
    }

    // Test nested async/await
    nestedAsync(): number {
        const inner = async () => {
            const value = await Promise.resolve(15);
            return value + 5;
        };
        return await inner(); // Should return 20
    }

    // Test conditional awaiting
    conditionalAwait(usePromise: boolean): number {
        if (usePromise) {
            return await Promise.resolve(100);
        } else {
            return 200;
        }
    }

    // Test exception handling with try/catch
    asyncTryCatch(): number {
        try {
            await Promise.reject(new Error("Test error"));
            return -1; // Should not reach this
        } catch (error) {
            return 50; // Should return this
        }
    }

    // Test finally block with async
    asyncFinally(): number {
        let result = 0;
        try {
            result = await Promise.resolve(10);
        } finally {
            result += 5;
        }
        return result; // Should return 15
    }

    // Test setTimeout-like delayed promise
    delayedPromise(delay: number, value: number): Promise<number> {
        return new Promise((resolve) => {
            // Simulate async delay (in real implementation this would use setTimeout)
            resolve(value);
        });
    }

    // Test concurrent operations
    concurrentOperations(): number {
        const promise1 = this.delayedPromise(100, 10);
        const promise2 = this.delayedPromise(200, 20);
        const promise3 = this.delayedPromise(50, 5);

        const results = await Promise.all([promise1, promise2, promise3]);
        return results[0] + results[1] + results[2]; // Should return 35
    }

    // Test promise constructor with immediate resolution
    immediateResolve(): number {
        const promise = new Promise((resolve, reject) => {
            const condition = true;
            if (condition) {
                resolve(123);
            } else {
                reject(new Error("Should not happen"));
            }
        });
        return await promise; // Should return 123
    }

    // Test promise constructor with immediate rejection
    immediateReject(): number {
        const promise = new Promise((resolve, reject) => {
            const condition = false;
            if (condition) {
                resolve(456);
            } else {
                reject(new Error("Immediate rejection"));
            }
        });
        return await promise; // Should throw exception
    }

    // Test promise with both resolve and reject paths
    conditionalPromise(shouldResolve: boolean): number {
        const promise = new Promise((resolve, reject) => {
            if (shouldResolve) {
                resolve(777);
            } else {
                reject(new Error("Conditional rejection"));
            }
        });
        return await promise;
    }

    // Test awaiting primitive values (should pass through)
    awaitPrimitives(): number {
        const num = await 42;
        const str = await "hello";
        const bool = await true;
        return num; // Should return 42
    }

    // Test mixed promise and non-promise values
    mixedValues(): number {
        const directValue = 10;
        const promiseValue = await Promise.resolve(20);
        const anotherDirect = 30;
        return directValue + promiseValue + anotherDirect; // Should return 60
    }

    // Test recursive async calls
    recursiveAsync(count: number): number {
        if (count <= 0) {
            return 0;
        }
        const currentValue = await Promise.resolve(count);
        const restValue = await this.recursiveAsync(count - 1);
        return currentValue + restValue;
    }

    // Test promise that resolves to another promise
    nestedPromiseResolution(): number {
        const innerPromise = Promise.resolve(88);
        const outerPromise = Promise.resolve(innerPromise);
        return await outerPromise; // Should return 88
    }

    // Test error propagation through promise chain
    errorPropagation(): number {
        return await Promise.resolve(10)
            .then(value => {
                if (value > 5) {
                    throw new Error("Value too large");
                }
                return value * 2;
            })
            .catch(error => 99); // Should return 99
    }
}
